File: java-scalacheck/src/main/scala/Account.scala
package com.company.ai.scala

/**
 * This file implements the same functionality as the equivalent Java class, but
 * implemented in Scala.
 *
 * Please note that the class being tested with ScalaCheck is the Java version, not
 * this one.
 */
case class InsufficientFundsException() extends Exception

object Account {
  val GOLD_AGE : Int = 50
  val GOLD_BALANCE : Double = 10000
  val STD_INTEREST : Double = .02
  val GOLD_INTEREST : Double = .03
}

case class Account(id: Int, var age: Int, private var _balance: Double) {
    // id and age are positive integers

    import Account._

    def balance = _balance

    def deposit(amt: Double) = {
        assert(amt > 0)
        _balance += amt
    }

    def withdraw(amt: Double) = {
        assert(amt > 0)
        if (amt <= _balance) _balance -= amt
        else throw InsufficientFundsException()
    }

    def rate = if (balance < GOLD_BALANCE && age < GOLD_AGE) STD_INTEREST
               else GOLD_INTEREST

    def creditInterest() = deposit(rate * balance)
}

File: java-scalacheck/src/main/com/company/account/Account.java
package com.company.account;

/**
 * Implementation of a simple bank account class
 */
public class Account {
  
  public final static int GOLD_AGE = 50;
  public final static double GOLD_BALANCE = 10000;
  public final static double STD_INTEREST = .02;
  public final static double GOLD_INTEREST = .03;
  
  private int id;
  private int age;
  private double balance;
  
  public Account(int id, int age, double balance) {
    this.id = id;
    this.age = age;
    this.balance = balance;
  }
  
  public double getBalance() {
    return(balance);
  }
  
  public int getAge() {
    return(age);
  }
  
  public void deposit(double amt) {
    assert (amt > 0);
    balance += amt;
  }
  
  public void withdraw(double amt) throws InsufficientFundsException {
    assert(amt > 0);
    if (amt <= this.balance) 
      balance -= amt;
    else 
     throw new InsufficientFundsException();
  }

  public double getRate() {
    if (balance < Account.GOLD_BALANCE && age < Account.GOLD_AGE) 
      return(Account.STD_INTEREST);
    else 
      return(Account.GOLD_INTEREST);
  }

    public void creditInterest() {
    deposit(getRate() * balance);
  }
}